File: 45-jump-game-ii/jump-game-ii.java
class Solution {
    public int jump(int[] nums) {
        int l=0,r=0,jump=0;
        while(r<nums.length-1){
            int far=Integer.MIN_VALUE;
            for(int i=l;i<=r;i++){
                far=Math.max(far,i+nums[i]);
            }
            l++;
            r=far;
            jump++;

        }
        return jump;
    }
}

File: 45-jump-game-ii/jump-game-ii.cpp
class Solution {
public:
    int solve(int ind, vector<int>& nums, vector<int>& dp) {
        if (ind == nums.size() - 1)
            return 0;
        if (ind >= nums.size())
            return 1e9;

        if (dp[ind] != -1)
            return dp[ind];

        int ans = 1e9;

        for (int jump = 1; jump <= nums[ind]; jump++) {
            ans = min(ans, 1 + solve(ind + jump, nums, dp));
        }

        return dp[ind] = ans;
    }

    int jump(vector<int>& nums) {
        vector<int> dp(nums.size(), -1);
        return solve(0, nums, dp);
    }
};